airbyte-integrations/connectors/source-shopify/source_shopify/shopify_graphql/bulk/query.py

```diff
@@ -0,0 +1,184 @@
+        """
+        Defines the root of the graph with edges.
+        """
+        args: List[Argument] = []
+        # constructing arguments
+        if filter_query:
+            args.append(Argument(name=self.operation, value=f'"{filter_query}"'))
+        if sort_key:
+            args.append(Argument(name="sortKey", value=sort_key))
+        # constructing edges
+        fields = [
+            Field(name=self.edge_key, fields=[Field(name=self.node_key, fields=["id", edges] if edges else ["id"])]),
+        ]
+        # return constucted query
+        return Query(name=name, arguments=args, fields=fields)
+
+
+class ShopifyBulkQuery(GraphQlQueryBuilder):
+    @property
+    @abstractmethod
+    def record_identifier(self) -> str:
+        """
+        Defines the record identifier to fetch only records related to the choosen stream.
+        Example:
+            { "admin_graphql_api_id": "gid://shopify/Metafield/22533588451517" }
+            In this example the record could be identified by it's reference = ".../Metafield/..."
+        The property should be defined like:
+            record_identifier = "Metafield"
+        """
+
+    @property
+    @abstractmethod
+    def graph_edge(self) -> str:
+        """
+        Defines the final graph node name to fetch from.
+        """
+
+    @property
+    @abstractmethod
+    def graph_fields(self) -> Union[List[str], List[Any]]:
+        """
+        Defines the fields for final graph selection.
+        """
+
+    @abstractmethod
+    def resolve_query(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_query: str = None,
+        sort_key: str = None,
+    ) -> Operation:
+        """
+        Defines how query object should be resolved based on the base query selection.
+        """
+
+    def __new__(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_field: str = None,
+        start: str = None,
+        end: str = None,
+        sort_key: str = None,
+    ) -> str:
+        if not query_path:
+            raise NotImplementedError("The `query_path` is not defined.")
+        else:
+            # define query string
+            filter_query = f"{filter_field}:>='{start}' AND {filter_field}:<='{end}'" if filter_field else None
+            return self.resolve_query(self, query_path, filter_query, sort_key)
+
+
+class Metafields(ShopifyBulkQuery):
+
+    record_identifier = "Metafield"
+    graph_edge = "metafields"
+
+    # list of available fieds:
+    # https://shopify.dev/docs/api/admin-graphql/unstable/objects/Metafield
+    graph_fields = [
+        "id",
+        "namespace",
+        "value",
+        "key",
+        "description",
+        "createdAt",
+        "updatedAt",
+        "type",
+    ]
+
+    def resolve_query(
+        self,
+        query_path: Union[List[str], str] = None,
+        filter_query: str = None,
+        sort_key: str = None,
+    ) -> Operation:
+        """
```

Comment:
Yeah, it's much easier to build something like GraphQL queries with this approach, but generally speaking, it still looks ugly to me, don't like GQL in general because of the query complexity.